src/abi/abi_type.ts

```diff
@@ -0,0 +1,846 @@
+  ByteLen() {
+    let size = 0;
+    for (let i = 0; i < this.childTypes.length; i++) {
+      if (this.childTypes[i].constructor === BoolType) {
+        const after = TupleType.findBoolLR(this.childTypes, i, 1);
+        const boolNum = after + 1;
+        i += after;
+        size += Math.trunc((boolNum + 7) / 8);
+      } else {
+        const childByteSize = this.childTypes[i].ByteLen();
+        size += childByteSize;
+      }
+    }
+    return size;
+  }
+
+  Encode(values: ABIValue[]) {
+    if (values.length > MAX_LEN) {
+      throw new Error('length of tuple array should not exceed a uint16');
+    }
+    const tupleTypes = this.childTypes;
+    const heads: Uint8Array[] = [];
+    const tails: Uint8Array[] = [];
+    const isDynamicIndex = new Map<number, boolean>();
+    let i = 0;
+
+    while (i < tupleTypes.length) {
+      const tupleType = tupleTypes[i];
+      if (tupleType.IsDynamic()) {
+        // Head is not pre-determined for dynamic types; store a placeholder for now
+        isDynamicIndex.set(heads.length, true);
+        heads.push(new Uint8Array([0, 0]));
+        tails.push(tupleType.Encode(values[i]));
+      } else {
+        if (tupleType.constructor === BoolType) {
+          const before = TupleType.findBoolLR(tupleTypes, i, -1);
+          let after = TupleType.findBoolLR(tupleTypes, i, 1);
+
+          // Pack bytes to heads and tails
+          if (before % 8 !== 0) {
+            throw new Error(
+              'expected before index should have number of bool mod 8 equal 0'
+            );
+          }
+          after = Math.min(7, after);
+          const compressedInt = TupleType.compressMultipleBool(
+            values.slice(i, i + after + 1)
+          );
+          heads.push(bigIntToBytes(compressedInt, 1));
+          i += after;
+        } else {
+          const encodedTupleValue = tupleType.Encode(values[i]);
+          heads.push(encodedTupleValue);
+        }
+        isDynamicIndex.set(i, false);
+        tails.push(new Uint8Array());
+      }
+      i += 1;
+    }
+
+    // Adjust head lengths for dynamic types
+    let headLength = 0;
+    for (const headElement of heads) {
+      headLength += headElement.length;
+    }
+
+    // Encode any placeholders for dynamic types
+    let tailLength = 0;
+    for (let j = 0; j < heads.length; j++) {
+      if (isDynamicIndex.get(j)) {
+        const headValue = headLength + tailLength;
+        if (headValue > MAX_LEN) {
+          throw new Error(
+            `byte length of ${headValue} should not exceed a uint16`
+          );
+        }
+        heads[j] = bigIntToBytes(headValue, LENGTH_ENCODE_BYTE_SIZE);
+      }
+      tailLength += tails[j].length;
+    }
+
+    // Concatenate into fixed Uint8Array
+    let resultBytes = new Uint8Array();
+    for (const h of heads) {
+      resultBytes = concatArrays(resultBytes, h);
+    }
+    for (const t of tails) {
+      resultBytes = concatArrays(resultBytes, t);
+    }
+    return resultBytes;
```

Comment:
nit: `return concatArrays(...heads, ...tails)` would be simpler

Reply:
Neat!